Adds authentication and unittests to OpenPubkey MFA cosigner protocol

EthanHeilman opened the pull request:
This PR builds on the MFA Cosigner introduced in [Add MFA Cosigner example #54
](https://github.com/openpubkey/openpubkey/pull/54)


https://github.com/openpubkey/openpubkey/assets/274814/852c720a-e686-4036-9d93-61804e3e4af6



This PR adds:
1. **Cosigner Client** - Adds cosigner support in the OpenPubkey client that allows a browser based OIDC auth flow to be redirected, on its conclusion, to a Cosigner auth flow.
2. **Authenticating Cosigner (AuthCosigner)**  - An authenticated cosigner which can be composed with an authentication method. This struct enforces an OpenPubkey Proof-of-Possession authentication check via an authcode and enables authentication session management via an AuthID. For flexibility it does not implement the user authentication check but can be composed with a user authentication check.
3. **MFA Cosigner example** - An example MFA Cosigner using Webauthn for user authentication is composed with the Authenticated Cosigner. This example is intended to prove the OpenPubkey core software has the needed functionality for MFA Cosigning and provide an example for to show how compose authentication methods with Authenticating Cosigner. The MFA Cosigner is not intended as an example and not a production WebAuthn cosigner.
4. **client.Auth** a new method on the client struct that calls OidcAuth and then if a cosigner is specified also performs cosigner auth. This is the default auth method on client. Future PRs may explore no longer exposing the OidcAuth method in client.
5. **Improved cosigner verification logic** Originally this was going to moved completely to PR #72 However after some experiments. I've decide move the verifier into this PR and use PR #72 to consider how to refactor the package restructure.
6. **Tests** including a webauthn device emulator that allows us to test the MFA Cosigner example.
7. **Copyright header:** To preserve git history when adding copyright headers, I've added the copyright header to all files changes by this PR. See Issue https://github.com/openpubkey/openpubkey/issues/82

## Things not done in this PR

- Full JWKS URI support via a composable JWKS server Struct
- MFA Authentication session expiration. Right now you can start an MFA session, go away for a week and then complete it.
- Full protocol documentation of MFACosigner (see https://github.com/openpubkey/openpubkey/issues/85). After this PR is merged the MFACosigner will have enough functionality locked down as to enable this level of documentation to be written.

## Cosigner Authentication Protocol

This protocol implements the PK Token Proof of Possession Authentication with the necessary session management keys to allow any MFA to be used with a cosigner. The user to Cosigner MFA authentication is left abstract to enable any authentication method to be used. To see how these abstract details can be filled in, we provide the webauthn based MFA Cosigner example.


```mermaid
sequenceDiagram

    participant user as User (Alice)
    box User's Machine
    participant opkc as OpenPubkey client
    participant wb as Browser window
    end
    participant cos as MFA Cosigner

    Note over opkc: OIDC auth flow finishes<br/>Cosigner auth begins
    
    wb->>opkc: OIDCAuth passes HTTP session to Cosigner client
    opkc->>wb: HTTP Redirect: PK Token, SIGN(usk, (time, nonce, rURI)
    wb->>cos: 
    Note over cos: Verifies signature,<br/> issues new AuthID for this auth session<br/> and associates it with sub ID in PK Token 
    
    cos->>wb: HTTP Redirect to user authn page: AuthID
    wb->>cos: Begin User Authn, AuthID
    cos->> wb: Authn challenge
    wb->>user: 
    user->>wb: User authn response, AuthID
    wb->>cos: 
    Note over cos: Authenticates user via authn response,<br/> and issues authcode 
    
    cos->>wb: Redirect (authcode)
    wb->>opkc: 
    opkc->>cos: authcode, Sign(usk, authcode)
    cos->>opkc: Issues Signature
```

1. **InitAuth  (OPK Client --> Cosigner):** OPK client creates the InitAuth message `InitMFAAuth ={TimeSigned: time.now, Nonce: RandomBytes(), RedirectUri: "http://localhost/mfacallback" }` signs it under their PK Token pkt to generate an OSM (OpenPubkey Signed Message) which we denote sig1. OPK Client sends pkt and sig1 to the Cosigner.

2. **Create AuthID (Cosigner)**: Cosigner verifies sig1 under pkt. If valid, it generates a new AuthID and stores a mapping between the AuthID and the identity in the PK Token.

3. **Issues and sends MFA Authentication Challenge (Cosigner-->User):**  The Cosigner issues an authentication challenge to the user and sends AuthID along with this challenge so that the response can be associated with the identity.

4. **Answers Authentication Challenge (User-->Cosigner):** The user answers the authentication challenge sending AuthID with the answer.

6. **Issues and sends authcode (Cosigner-->OPK Client):** Now that the user has authenticated the Cosigner creates an authcode associated with the AuthID. Cosigner sends the authcode to the OPK Client using the RedirectURI specified in the InitAuthCode message.

7. **Signature Request (OPK Client-->Cosigner):** OPK Client signs the authcode under their PK Token, pkt, to create an OSM (OpenPubkey Signed Message) sig2 and sends signed sig2 and their pkt to the Cosigner.

8. **Signature Issuance (Cosigner-->OPK Client):** Cosigner checks that sig2 verifies under the pkt PK Token sent and then uses the authcode in sig2 to look up the AuthID for this session. It then verifies the PK Token associated with the AuthID is the same as PK Token send in the Signature Request. The cosigner then cosigns this PK Token. The Cosigner includes the RedirectURI and the Nonce as fields in the cosigner claims which are stored in the protected header of the signature. and sends the signature and cosigner claims to the OPK Client.

9. **Verification of Cosigner Signature: (OPK Client):** OPK Client checks that returned signature and cosigner claims are valid for the pkt. 
    Checking that:

    * RedirectURI specified in the Cosigner claims is the expected  RedirectURI (as set in the InitAuthMsg)
    * Nonce specified in the Cosigner claims is the expected Nonce (as set in the InitAuthMsg)
    * That the issuer specified in the Cosigner claims is the expected issuer
    * That the PK Token verifies with the Cosigner signature and claims (Downloads Cosigner public key based on issuer)
    
    If any of these conditions are not met, the OPK Client deletes the Cosigners Signature and outputs an error message.

The Cosigner Authentication Protocol is not specific to any particular authentication factor and the implementation of the Cosigner Authentication Protocol leaves the Authentication challenge and answer undefined.  This protocol is general to any sort of authentication factor (password, geographical location, IP address, etc...). To specify the exact authentication factors you wish to use, an implementor must compose the AuthCosigner struct with a struct that implements an authentication protocol. To see how this is done we provide an example Webauthn Cosigner which is composed with AuthCosigner.

### Why sign InitAuthMsg?

The careful reader will notice that since we perform a Proof-of-Possession with the Cosigner to redeem the authcode for the cosigners signature on the PK Token, the signature on the InitAuthMsg is not strictly needed to authenticate the user to the Cosigner.  Why sign the InitAuthMsg then? 
1.  The signature over the InitAuthMsg is performed to prevent someone how knows a users PK Token but not their signing key from checking if the user has registered an authentication factor with the cosigner. Requiring authentication before issuing an MFA challenge is standard practice. 
2. It also shields an attacker from attempting to brute force the MFA factor in cases where the MFA factor has limited entropy: ToTP, etc...

### Why include a nonce?

The nonce enables the user's client to be sure the authentication session it initiated is the same authentication session it completed. This rules out the possibility of an attacker swapping in InitMFAAuth messages. I am not aware of a concrete attack here that would violate security, but I did not feel comfortable not having a mechanism for a user's client to prevent the substitution of one authentication session for another one.

This nonce field is inspired by [the nonce field in OpenID Connect.](https://openid.net/specs/openid-connect-core-1_0.html#NonceNotes)

## OPK Client Cosigner support

This PR Cosigner support in the OPK client by the creation of a Cosigner provider which enables a client using a web browser based OIDC auth flow to reuse the browser window for the Cosigner auth flow. To do this, we add an HTTP Server session hook to the OIDC provider so that when the OIDC auth flow completes, the Cosigner auth flow can redirect the user's open browser window to the Cosigner URI.

As a result of this, the localhost webserver created by the OIDC Provider to receive the authcode can not be closed when `OP.RequestTokens` completes. Instead, if session hook is set, it must now wait until the Cosigner provider redirects the users browser window. To ensure that when the session hook is provided we shutdown the localhost webserver we add a shutdown call in both the err channel for the OIDC Provider and the callback http handler. Thus, on success we shutdown the http server and on failure we shutdown the http server.

### Why a separate http server for the Cosigner Provider in the client?

The Cosigner provider opens it's own HTTP server to receive the authcode from the Cosigner. We prefer creating a new HTTP server rather than reusing the HTTP server created in the OIDC provider for the following reasons:

1. It enables flexibility in that the Cosigner provider can use a different port than the OIDC Provider.  Consider the weirdness of requiring that the OIDC provider and the Cosigner must always use the same port. 
2. It lets the Cosigner provider be used independently of the OIDC Provider, allowing the Cosigner client to eventually support an OIDC Provider which isn't using a web based flow.
3. It makes shutdown of the OIDC provider's http server far easier.

## Example MFA Cosigner (using the Cosigner Authentication Protocol)

This PR also includes a MFA Cosigner built on webauthn to demonstrate how an MFA cosigner can be built from the Authenticating Cosigner.

```mermaid
sequenceDiagram

    participant user as User (Alice)
    box User's Machine
    participant opkc as OpenPubkey client
    participant wb as Browser window
    end
    participant cos as MFA Cosigner

    Note over opkc: OIDC auth flow finishes<br/>Cosigner auth begins
    
    wb->>opkc: OIDCAuth passes HTTP session to Cosigner client
    opkc->>wb: HTTP Redirect: PK Token, SIGN(usk, (time, nonce, rURI)
    wb->>cos: 
    Note over cos: Verifies signature,<br/> issues new AuthID for this auth session<br/> and associates it with sub ID in PK Token 
    
    cos->>wb: AuthID
    wb->>cos: Webauthn Begin Login, AuthID
    cos->>wb: Webauthn Challenge
    wb->>user: 
    user->>wb: User authentication response, AuthID
    wb->>cos: 
    Note over cos: Authenticates user via Webauthn device,<br/> and issues authcode 
    cos->>wb: Redirect (authcode)
    wb->>opkc: 
    opkc->>cos: authcode, Sign(usk, authcode)
    cos->>opkc: Issues Signature
```

## PKToken Verification struct

Prior to this PR when verifying a PK Token, we check that OP Issuer and OP Audience in the ID Token matches the issuer and audience in the Provider struct and the OP and Cosigner signature validates. The current approach has three limitations:

1. Only one OP can be provided to `client.VerifyPKToken(cxt, pkt, OP)`. This means if a verifying party wanted to accept two or more OPs, `op1, op2, ... opN1`, they would have call: `client.VerifyPKToken(cxt, pkt, op1)`, ... `client.VerifyPKToken(cxt, pkt, opN)`. Each of these would involve downloading each of the OP public keys and doing signature validation.

2. More of an issue is that `client.VerifyPKToken` does not provide a parameter to supply the allowed cosigners. As such any cosigner will be accepted. This logic prevents the cosigner signature from providing any security value.

3. If PK Token does not include a Cosigner signature, then it just automatically bypasses the cosigner verification. This is because the existing code does not have a way for a relying party to signal that a cosigner is or is not required.

We solve both of these problems by creating a PKTokenVerifier struct which allows specifying an allowlist of OPs and an allowlist of cosigners. This allowlist is checked against the PK Token before any remote JWKS calls are made to OPs or Cosigners.

```golang
verifier := client.PKTokenVerifer{
	AllowedProviders: []client.OpenIdProvider{provider},
	AllowedCosigners: []client.CosignerProvider{cosignerProvider},
}
err := verifier.Verify(context.TODO(), pkt)
```

If a Cosigner requirement is not desired then that is signaled by not setting the AllowedCosigners list.

```golang
verifier := client.PKTokenVerifer{
	AllowedProviders: []client.OpenIdProvider{provider},
}
err := verifier.Verify(context.TODO(), pkt)
```

To make this PR easier to review and to avoid merge conflicts with other active PRs we add the verifier to the verification code in `oidc.go.` A case could be made that since `oidc.go` is not correctly named as it contains more than just oidc code for the client. It is worth thinking about renaming this file or pulling the verification code into a different package or file in future PRs.

Originally this change was in https://github.com/openpubkey/openpubkey/pull/72 but I simplified it and pulled it into this PR.

**Known Limitations of the verifier struct as written**

* **Multiple providers with the same issuer and different audiences.** This does not handle the case where the same provider (OP) is specified twice in an allowlist with different audiences. Instead it will match on the first provider with the correct issuer and then verify using that provider. If that provider has the wrong audience it will throw an error even if there was a provider in the allowlist that had both the correct audience and the correct issuer. This limitation results from the fact that not all providers care about matching on the audience. It seems unlikely that user's would want to add the same issuer twice to the allowlist. See https://github.com/openpubkey/openpubkey/issues/84

* **Does not provide a way of pairing cosigners and providers.** A user may wish to only use a particular cosigner with a particular OP. This verifier does not provide a way of representing that relationship. A user would have to create multiple verifiers. It is unclear if any users want this feature and adding this capability would make configuring the verifier more complex for a user so we opted against adding it.



## Notes

* This PR does not include https://github.com/openpubkey/openpubkey/pull/63 so if you attempt to test this PR using the google example rather then MFA example you will run into the bug #63 solves. This should not prevent merging as this bugfix is automatically included in that merge operation.
~~* I renamed RedirURI to RedirectURI. Not sure about this, might change it back to touch less code.~~
* Tested this on windows and linux. Additionally on Linux I ran the race condition tester `go test -race ./...`. See results below

<img width="723" alt="Screen Shot 2024-01-08 at 6 03 00 PM" src="https://github.com/openpubkey/openpubkey/assets/274814/54a6b0df-4728-4487-a6a0-620c8b0ac7f1">

ymarcus93 reviewed (commented):
Really cool! Have not tested the example yet but I've reviewed all the code

EthanHeilman commented on client/oidc.go:
```diff
@@ -23,9 +24,49 @@ type OpenIdProvider interface {
+		return fmt.Errorf("no allowed providers (allowed OPs) set, you must set at least one provider")
+	}
+
+	var pktOp struct {
+		Issuer string `json:"iss"`
+	}
+	if err := json.Unmarshal(pkt.Payload, &pktOp); err != nil {
+		return err
+	}
+	var provider OpenIdProvider
+	for _, allowedOp := range v.AllowedProviders {
+		if allowedOp.GetIssuer() == pktOp.Issuer {
+			provider = allowedOp
+			break
+		}
+	}
+	if provider == nil {
+		return fmt.Errorf("the OP issuer (%s) in the PK Token is not an allowed issuer", pktOp.Issuer)
+	}
```
A comment makes sense here. I eventually want to enable the use of the same issuer with different audiences, I'm not sure the best way to do that since audience isn't exposed by the OP (provider) struct and even if it was, we would need to be careful not to try to match it for OPs that use the audience for the CIC.

Probably the ideal solution is a function on the OP interface that allows each OP struct to check if a PK Token "matches" without doing the full round trip to the JWKS endpoint.

EthanHeilman replied:
I created an issue for this and added a comment to document it. As this PR touches too much code, I plan to address this in a future PR

https://github.com/openpubkey/openpubkey/issues/84

EthanHeilman commented on client/oidc.go:
```diff
@@ -80,9 +121,30 @@ func VerifyPKToken(ctx context.Context, pkt *pktoken.PKToken, provider OpenIdPro
+				return fmt.Errorf("the COS issuer (%s) in the PK Token is not an list of allowed issuers", cosIss)
+			}
+		}
+		if pkt.Cos != nil {
+			if err := pkt.VerifyCosignerSignature(); err != nil {
+				return fmt.Errorf("error verify cosigner signature on PK Token: %w", err)
+			}
```
`if pkt.Cos != nil {` shouldn't be there. It is still secure because on line 126 I check `if pkt.Cos != nil {` and fail if it is nil so the impact here is just a NOP, but it should be removed.

EthanHeilman replied:
Fixed

EthanHeilman commented on client/client.go:
```diff
@@ -16,7 +17,41 @@ import (
 type OpkClient struct {
-	Op OpenIdProvider
+	Op   OpenIdProvider
+	CosP *CosignerProvider
```
Marking this as won't fix.

To cut down on interface pollution I avoid creating interfaces until I have more than two examples of a struct. Given that cosigner behavior is strictly defined and we generally don't expect developers to write their own CosignerProvider implementations (we do expect them to write their own Cosigner implementations but not Cosigner Provider). I don't see the benefit of adding an interface here. 

That said, once we get a sense what a generalized behavior an cosigner provider should provide and we see the need for more than on implementation of a cosigner provider, we can decide to replace this with an interface.

ymarcus93 replied:
Yeah I guess I was thinking in terms of unit testing strictly the logic in the `OpkClient` struct, where `CosP` is used, _without_ having to spin up a Cosigner implementation test http server and have the `CosignerProvider` talk to it (which also makes an http server for the callback path) in order to test `OpkClient`; an interface would solve this. 

But that may not be warranted just yet because it looks like `OpkClient` doesn't have too much extra logic besides just calling methods on `Op` and `CosP`.

Side note though: for an integration test of Cosigner logic, we can def do something like that where we spin up a test http server for those endpoints.

EthanHeilman commented on client/oidc.go:
```diff
@@ -23,9 +24,49 @@ type OpenIdProvider interface {
+	}.Verify(ctx, pkt)
+}
+
+type PKTokenVerifer struct {
```
Fixed

EthanHeilman commented on client/cos.go:
```diff
@@ -0,0 +1,181 @@
+
+	pktJson, err := json.Marshal(pkt)
+	if err != nil {
+		return nil, fmt.Errorf("cosigner client hit error serializing PK Token: %w\n", err)
```
Fixed

EthanHeilman commented on client/cos.go:
```diff
@@ -0,0 +1,181 @@
+	// This is where we get the authcode from the Cosigner
+	mux.Handle(c.CallbackPath,
+		http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+
+			// Get authcode from Cosigner via Cosigner redirecting user's browser window
+			params := r.URL.Query()
+			if _, ok := params["authcode"]; !ok {
+				errCh <- fmt.Errorf("Cosigner did not return an authcode in the URI")
+				return
+			}
+			authcode := params["authcode"][0] // This is the authcode issued by the cosigner not the OP
+
+			// Sign authcode from cosigner under PK Token and send signed authcode to Cosigner
+			sig2, err := pkt.NewSignedMessage([]byte(authcode), signer)
+			if err != nil {
+				errCh <- err
+				return
+			}
+			res, err := http.Get(fmt.Sprintf("%s/sign?sig2=%s", c.Issuer, sig2))
+			if err != nil {
+				errCh <- fmt.Errorf("error requesting MFA cosigner signature: %w", err)
+				return
+			}
+
+			// Receive response from Cosigner that has cosigner signature on PK Token
+			resBody, err := io.ReadAll(res.Body)
+			if err != nil {
+				errCh <- fmt.Errorf("error reading MFA cosigner signature response: %w", err)
+				return
+			}
+			cosSig, err := util.Base64DecodeForJWT(resBody)
+			if err != nil {
+				errCh <- fmt.Errorf("error reading MFA cosigner signature response: %w", err)
+				return
+			}
+			// Success
+			w.Write([]byte("You may now close this window"))
+			ch <- cosSig
+		}),
+	)
```
AFAICT you can add new handlers to a mux on the fly since the mux just manages a table of handlers. That said, I agree with you that it is nicer to add the handler first and then run the server to avoid getting a hard to debug 404 if the server responds to a request before the handler is added.

EthanHeilman replied:
Fixed

EthanHeilman commented on client/oidc.go:
```diff
@@ -23,9 +24,49 @@ type OpenIdProvider interface {
+	OpenIdProvider
+	HookHTTPSession(h HttpSessionHook)
+}
+type HttpSessionHook func(w http.ResponseWriter, r *http.Request)
```
Good catch, fixed

EthanHeilman commented on client/cos.go:
```diff
@@ -0,0 +1,181 @@
+	}
+
+	// Trigger redirect of user's browser window to a URI controlled by the Cosigner sending the PK Token in the URI
+	redirCh <- fmt.Sprintf("%s/mfa-auth-init?pkt=%s&sig1=%s", c.Issuer, string(pktB63), string(sig1))
```
Addressed, that is nicer

EthanHeilman commented on client/cos.go:
```diff
@@ -0,0 +1,181 @@
+		pkt.AddSignature(cosSig, pktoken.Cos)
+		if err != nil {
+			return nil, fmt.Errorf("error in adding cosigner signature to PK Token: %w\n", err)
+		}
```
Fixed

EthanHeilman commented on client/cos.go:
```diff
@@ -0,0 +1,181 @@
+		return pkt, nil
+	case err := <-errCh:
+		return nil, err
+	}
```
Fixed

EthanHeilman commented on client/cos.go:
```diff
@@ -0,0 +1,181 @@
+	}()
+	defer server.Shutdown(ctx)
+
+	ch := make(chan []byte)
+	errCh := make(chan error)
```
Let's see if I understand what you are saying:

`errCh := make(chan error)` is equivalent to `errCh := make(chan error, 0)`. This means if an error occursin the CallbackPath-handler we before we reach `case err := <-errCh:` then the CallbackPath-handler function be blocked. Then if `ctx.Done()` is called, we might never reach `case err := <-errCh:` and CallbackPath-handler will block indefinitely (or at least until process tear down). 

Where as if 
```golang
ch := make(chan []byte, 1)
errCh := make(chan error, 1)
```
CallbackPath-handler will not block and always execute the return statement following each `ch` or `errCh` send?

EthanHeilman replied:
Fixed to have 1 buffer channels

ymarcus93 replied:
> Let's see if I understand what you are saying:
> ...
> Then if ctx.Done() is called, we might never reach case err := <-errCh: and CallbackPath-handler will block indefinitely (or at least until process tear down).

> CallbackPath-handler will not block and always execute the return statement following each ch or errCh send?

Yes that was my original concern. The buffer of size `1` definitely helps alleviate the issue, but on second thought, I guess technically there is still a very small chance for that handler to still be blocking in case two errors are received before `server.Shutdown()` can be called after the `ctx` is cancelled. So the buffer size only makes the issue less likely. Maybe this isn't too much of a concern though--let's just keep it with buffer of size 1 for now.

EthanHeilman commented on client/cos.go:
```diff
@@ -0,0 +1,181 @@
+				errCh <- err
+				return
+			}
+			res, err := http.Get(fmt.Sprintf("%s/sign?sig2=%s", c.Issuer, sig2))
```
Fixed

EthanHeilman commented on client/cos.go:
```diff
@@ -0,0 +1,181 @@
+	}
+
+	// Trigger redirect of user's browser window to a URI controlled by the Cosigner sending the PK Token in the URI
+	redirCh <- fmt.Sprintf("%s/mfa-auth-init?pkt=%s&sig1=%s", c.Issuer, string(pktB63), string(sig1))
```
Fixed

EthanHeilman commented on cosigner/cosigner.go:
```diff
@@ -0,0 +1,26 @@
+
+type Cosigner struct {
+	alg    jwa.KeyAlgorithm
+	signer crypto.Signer
```
Fixed by exposing Alg and Signer.

EthanHeilman commented on cosigner/authcosigner.go:
```diff
@@ -0,0 +1,224 @@
+		}
+		audience = strings.Join(audList, ",")
+	default:
+		return nil, fmt.Errorf("failed to deserialize aud (audience) claim in ID Token: %d", t)
```
Yeah `%T` is the correct format string for this purpose, fixed by switching to `%T`

EthanHeilman commented on cosigner/authcosigner.go:
```diff
@@ -0,0 +1,224 @@
+	}
+}
+
+func (c *AuthCosigner) CreateAuthID(timeNow uint64) (string, error) {
+	authIdInt := c.authIdIter.Add(1)
+	iterAndTime := []byte{}
+	iterAndTime = binary.LittleEndian.AppendUint64(iterAndTime, uint64(authIdInt))
+	iterAndTime = binary.LittleEndian.AppendUint64(iterAndTime, timeNow)
+	mac := hmac.New(crypto.SHA3_256.New, c.hmacKey)
+	if n, err := mac.Write(iterAndTime); err != nil {
+		return "", err
+	} else if n != 16 {
+		return "", fmt.Errorf("unexpected number of bytes read by HMAC, expected 16, got %d", n)
+	} else {
+		return hex.EncodeToString(mac.Sum(nil)), nil
+	}
+}
```
I avoided using a UUID here because they only have 122-bits, so under the birthday bound, they only have 64-bits of collision resistance. As [wikipedia says](https://en.wikipedia.org/wiki/Universally_unique_identifier) to get a 50% chance of collision with a UUIDv4 you would need to generate "1 billion UUIDs per second for about 86 years", so 122-bits is probably safe.

Then again [the OAuth2 spec](https://datatracker.ietf.org/doc/html/rfc6749#section-10.10) hates on the idea of anything less than 128-bit collision resistance:

>  The probability of an attacker guessing generated tokens (and other credentials not intended for handling by end-users) MUST be less than or equal to 2^(-128) and SHOULD be less than or equal to 2^(-160).

I just didn't want to have to have to justify why 64-bit is "good enough" in this case. So I choose a 256-bit (16-byte) ID giving me 128-bits of collision resistance. Most of OpenPubkey is >=256-bit values so anyone doing a security analysis so it aligns well with the other cryptographic values.

EthanHeilman commented on cosigner/authcosigner.go:
```diff
@@ -0,0 +1,224 @@
+	KeyID        string
+	authIdIter   atomic.Uint64
+	hmacKey      []byte
+	AuthStateMap map[string]*AuthState
+	AuthCodeMap  map[string]string
```
For `AuthCodeMap` we only write to it once and the key we use is randomly chosen so

EthanHeilman replied:
> This assumption stems from developers viewing the table[key] syntax used for map accesses and misinterpreting them to be accessing disjoint elements. However, a map (hash table), unlike an array or a slice, is a sparse data structure, and accessing one element might result in accessing another element; if during the same process another insertion/deletion happens, it will modify the sparse data structure and cause a data race

maps in go are gross

EthanHeilman replied:
Looking into this more, I am just going to use RWMutex around AuthStateMap and AuthCodeMap. Anyone using this an actual MFACosigner should replace AuthStateMap and AuthCodeMap with a database.

EthanHeilman replied:
Fixed, but doublecheck my locking code.

I ended up completely rewriting this by adding a AuthStateStore interface and then creating a inmemory AuthStateStore which has locks on both maps.

I want to revisit this after we have a database backed production ready Cosigner to ensure the AuthStore interface I designed here actually exposes the right functions. Should it also expose the user or session maps in the example MFACosigner? I don't know and it is hard to know without a good database backed implementation.

ymarcus93 replied:
Locking code lgtm. I think if you run with the `-race` flag when running unit tests, as long as the unit tests touch code that involve the maps, we will find out if there are any race conditions still

ymarcus93 replied:
> I want to revisit this after we have a database backed production ready Cosigner to ensure the AuthStore interface I designed here actually exposes the right functions. Should it also expose the user or session maps in the example MFACosigner? I don't know and it is hard to know without a good database backed implementation.

Yeah let's wait for now. Seems good to me right now though. It might slightly change later if we think we need to expose some more things

EthanHeilman commented on cosigner/authcosigner_test.go:
```diff
@@ -0,0 +1,252 @@
+	authID1, err := cos.CreateAuthID(unixTime)
+	if err != nil {
+		t.Fatal(err)
+	}
```
Fixed!

EthanHeilman commented on cosigner/authcosigner.go:
```diff
@@ -0,0 +1,224 @@
+func (c *AuthCosigner) InitAuth(pkt *pktoken.PKToken, sig []byte) (string, error) {
+	msg, err := pkt.VerifySignedMessage(sig)
+	if err != nil {
+		return "", fmt.Errorf("failed to parse sig: %w", err)
```
Fixed!

EthanHeilman commented on cosigner/authcosigner_test.go:
```diff
@@ -0,0 +1,252 @@
+	if err != nil {
+		t.Error(err)
+	}
+	require.NotEmpty(t, authID1)
```
Based on the changes detailed below I am marking this as fixed. Open a new comment if there are still changes you want here.

> I only ask because it makes this unit test more isolated to the cosigner package if we don't have to use the client package in order to test AuthCosigner.

This test is doing double duty right now, testing both AuthCosigner and CosignerProvider message construction. Longer term I'd like to fully isolate them, but I don't think it is worth it until we have Cosign functionality locked down.

CallbackPath is used as part of the redirectURI sent the the Cosigner. Based on your comment I have added the following check to CreateInitAuthSig

```golang
	if !strings.HasSuffix(redirectURI, c.CallbackPath) {
		return nil, "", fmt.Errorf("redirectURI (%s) does not end in expected callbackPath (%s)", redirectURI, c.CallbackPath)
	}
```

Regarding the issuer issue: thinking about it more, I've come to the conclusion that the issuer should be in the signed message and I've made that change.

```golang
type InitMFAAuth struct {
	Issuer      string `json:"iss"`
	RedirectUri string `json:"ruri"`
	TimeSigned  int64  `json:"time"`
	Nonce       string `json:"nonce"`
}
```

This rules out cross-issuer replay attacks since the cosigner will discard a signed message that doesn't include the issuer. Your comment was very helpful because once after adding this check on the issuer a bunch of tests failed where I had the wrong issuer. It is a good sanity check to have and improves security.

EthanHeilman commented on cosigner/authcosigner_test.go:
```diff
@@ -0,0 +1,252 @@
+		hmacKey:      hmacKey,
+		AuthStateMap: make(map[string]*AuthState),
+		AuthCodeMap:  make(map[string]string),
+	}
```
Fixed!

EthanHeilman commented on examples/mfa/mfacosigner/server.go:
```diff
@@ -0,0 +1,248 @@
+
+	mux := http.NewServeMux()
+	mux.Handle("/", http.FileServer(http.Dir("mfacosigner/static")))
+	mux.HandleFunc("/mfa-auth-init", server.initAuth)
+	mux.HandleFunc("/check-registration", server.checkIfRegistered)
+	mux.HandleFunc("/register/begin", server.beginRegistration)
+	mux.HandleFunc("/register/finish", server.finishRegistration)
+	mux.HandleFunc("/login/begin", server.beginLogin)
+	mux.HandleFunc("/login/finish", server.finishLogin)
+	mux.HandleFunc("/sign", server.signPkt)
```
Agreed, originally the auth cosigner was built using the OpenPubkey paper as the spec, but it evolved beyond that. 

I created this issue to track this

https://github.com/openpubkey/openpubkey/issues/85

lgmugnier replied:
I think this would be important to include in this PR. If you're creating something that people need to integrate against, it's important to tell them what they need to build for the integration. This could be accomplished with an interface.

EthanHeilman replied:
> This could be accomplished with an interface.

If they are implementing a cosigner in golang, they should use the provided client and server code in `client` and `/cosigner` rather than writing this own from scratch. However that only applies to the AuthCosigner.

Is the interface discussion relevant to `examples/mfa/mfacosigner/server.go`?

EthanHeilman commented:
@ymarcus93 Requesting a review. I believe I resolved or addressed all the issues you brought up with the exception create an interface for the cosigner provider.

I think we need a cosigner provider interface, I'm just not ready to add it to the PR yet.

ymarcus93 commented on cosigner/authstatestores.go:
```diff
@@ -0,0 +1,168 @@
+	s.AuthStateMapLock.RLock()
+	as, ok := s.AuthStateMap[authID]
+	s.AuthStateMapLock.RUnlock()
+	return *as, ok // Pass by value to prevent writes to the original
```
I think I get what you're trying to achieve by making `LookupAuthState` and `RedeemAuthcode` return values instead of pointers (to make a copy and prevent writes to the original value backed by the `map`), but since `AuthState` has a field that is a pointer (`Pkt`), it's not truly a copy (only the fields not backed by pointers are copied). This is really a shallow copy behind the scenes--the returned `AuthState` will have a copy of the `Pkt` pointer to the original `Pkt` of the `AuthState` stored in the `map`.

EthanHeilman replied:
This has been fixed

ymarcus93 commented on client/cos.go:
```diff
@@ -0,0 +1,229 @@
+
+	select {
+	// Trigger redirect of user's browser window to a URI controlled by the Cosigner sending the PK Token in the URI
+	// case redirCh <- fmt.Sprintf("%s/mfa-auth-init?pkt=%s&sig1=%s", c.Issuer, string(pktB63), string(sig1)):
```
leftover comment

EthanHeilman replied:
Fixed

EthanHeilman commented:
@ymarcus93 Not sure why github won't let me respond to your comment:

> The buffer of size 1 definitely helps alleviate the issue, but on second thought, I guess technically there is still a very small chance for that handler to still be blocking in case two errors are received before server.Shutdown() can be called after the ctx is cancelled. So the buffer size only makes the issue less likely. Maybe this isn't too much of a concern though--let's just keep it with buffer of size 1 for now.

My reasoning is as follows:
1. Each error sent to the error channel `errCh` immediately calls return. Since `errCh` is added by a closure, once return has been called no other errors may be sent. 
2. The channel `ch` is only called once at the end of the function so there can only be zero or one message sent via `ch`.

As such it should never block. If can think of a way it will block let me know because it means I made a mistake.

EthanHeilman commented:
@ymarcus93  Successfully passed with `go test -race ./...` on ubuntu. See PR for screenshot

lgmugnier commented on client/cos.go:
```diff
@@ -0,0 +1,261 @@
+
+func (c *CosignerProvider) RequestToken(ctx context.Context, signer crypto.Signer, pkt *pktoken.PKToken, redirCh chan string) (*pktoken.PKToken, error) {
+	// Find an unused port
+	listener, err := net.Listen("tcp", ":0")
+	if err != nil {
+		return nil, fmt.Errorf("failed to bind to an available port: %w", err)
+	}
```
I think anything with a server in it likely belongs in the examples folder. Our core code should be proposing a framework with specific parts filled out and others left blank for customers to plug into. I think that was a big part of the V1 architecture. Here, I think we're assuming too much and aren't building in the flexibility this kind of library should have.

EthanHeilman replied:
This is following the HTTP server pattern we already use in the client to get the callback from the OP. 

```golang
	logrus.Infof("listening on http://%s/", lis)
	logrus.Info("press ctrl+c to stop")
	earl := fmt.Sprintf("http://localhost:%s/login", g.RedirURIPort)
	util.OpenUrl(earl)

	go func() {
		err := g.server.ListenAndServe()
		if err != nil && err != http.ErrServerClosed {
			logrus.Error(err)
		}
	}()
```
https://github.com/openpubkey/openpubkey/blob/main/client/providers/google.go#L92

lgmugnier replied:
Yup, I also think that should be in examples but I was just making a comment on the large amount of new http code in this PR.

lgmugnier commented on client/oidc.go:
```diff
@@ -1,3 +1,19 @@
+//
+// SPDX-License-Identifier: Apache-2.0
+
 package client
```
I think this deserves its own package: `validator`. Helps with distinction architecturally and organizationally.

EthanHeilman replied:
I agree. I want to move validating into the a validator package, but I experimented with this in some draft PRs. The main issue I ran into is the cyclical dependencies between the client and providers that are exposed when client depends on the validator package. My next PR after this PR is fixing the cyclical dependencies in the client, which should enable creating a nice validator package.

lgmugnier commented on cosigner/authcosigner.go:
```diff
@@ -0,0 +1,117 @@
+	}, nil
+}
+
+func (c *AuthCosigner) InitAuth(pkt *pktoken.PKToken, sig []byte) (string, error) {
```
It looks like a lot of complex code is being added to support this version of PoP Auth. I strongly believe that this is something we should add after we have fully ratified a design. Especially since it's not a requirement of V1, I don't see a reason to rush it.

EthanHeilman replied:
This is just to authenticate the user to the cosigner. It is not the full POP Auth protocol from the paper as that requires Refresh and is general purpose.

lgmugnier commented on cosigner/cosigner.go:
```diff
@@ -0,0 +1,42 @@
+	"github.com/openpubkey/openpubkey/pktoken"
+)
+
+type Cosigner struct {
+	Alg    jwa.KeyAlgorithm
+	Signer crypto.Signer
+}
```
Wait, we have a `Cosigner` and a `CosignerProvider`? Can you explain why we need both?

EthanHeilman replied:
Cosigner and AuthCosigner which does the Cosigning, CosignerProvider is the client which talks to the Cosigner. I debated CosignerProvider vs CosignerClient, but given that we call the OpenIDClient the OpenIDProvider it seemed like a more natural name.

lgmugnier commented on examples/mfa/example.go:
```diff
@@ -41,62 +49,66 @@ func main() {
 
-	opk := &client.OpkClient{
-		Op: provider,
+	cosignerProvider := client.CosignerProvider{
```
Am I missing something, is `client.CosignerProvider` getting instantiated somewhere above? Otherwise I'm pretty sure this will throw an error

EthanHeilman replied:
I might be misunderstanding you. `client` is the name of the package, so `client.CosignerProvider{...}` is creating a new CosignerProvider struct.

lgmugnier replied:
oh right, yeah I misunderstood thanks!

lgmugnier commented on client/oidc.go:
```diff
@@ -76,7 +92,54 @@ type OpenIdProvider interface {
 	VerifyNonGQSig(ctx context.Context, idt []byte, expectedNonce string) error
 }
 
+// Interface for interacting with the OP (OpenID Provider)
+type BrowserOpenIdProvider interface {
+	OpenIdProvider
+	HookHTTPSession(h http.HandlerFunc)
+}
```
I don't think this interface belongs in this file?

EthanHeilman replied:
What do you suggest we move it to?

lgmugnier replied:
Typically it goes closest to where it's used so likely `client.go`. I just didn't know if you had put it in this file for some reason that I missed.

EthanHeilman replied:
I put it in oidc.go since it is an OIDC provider and it makes the most sense to me to put is directly below the `OpenIDProvider` interface definition which it extends. That said, it also makes sense to put in it `client.go`, I had it there at point. I have no objection to moving it to `client.go` if you want me to move it.

EthanHeilman commented:
> I agree with yuval that we need an interface here, otherwise how are people supposed to plug in their own?

@lgmugnier Assuming the other parties are using the MFA Cosigner protocol they should just be able to configure the existing CosignerProvider without having to crate a new one. As much as possible I want to encourage people to use the existing CosignerProvider. No sense creating an interface until we need it.

I do agree that we will likely want an CosignerProvider interface at some point but I don't want to create one until we have a concrete example a second implementation.
 
@ymarcus93 mentioned adding a CosignerProvider for testing. I have no objection to adding an interface in a PR that added a testing second CosignerProvider which needed an interface here.

EthanHeilman commented on client/client.go:
```diff
@@ -17,7 +34,41 @@ import (
+
+// Auth will attempt to authenticate to both the OpenID Provider and then the Cosigner.
+// If no Cosigner supplied it will default to OidcAuth
+func (o *OpkClient) Auth(
```
Thats a really neat idea. If we can get it to work I think it would be a great improvement.

How does the redirect `CosignPkToken` trigger the redirect inside of `GetPKToken`? How does `GetPKToken` know to keep the HTTP session open so that the redir can be triggered?

lgmugnier replied:
What I'm proposing also assumes that that code also has control over the server (I've made some comments about this elsewhere), so that people can plug in their own.

The redirect would happen as soon as the first call is done, and the triggering wouldn't necessarily require the HTTP session being left open, you can accomplish it with web page code too and a signal from the local server.

EthanHeilman replied:
@johncmerfeld  

Regarding `client.GetPKToken(...)` and `client.CosignPkToken( ...)` I'm not sure how this would work. I'm not saying is can't work but I don't see how to do it without opening a second browser window. Another issue is that depending on how the client is configured it may or may not need to do the redirect. I don't see how `client.GetPKToken(...) ` would know if it should do a redirect or not unless it was previously configured to expect the user to call `client.CosignPkToken( ...)`, but if the configuation makes calling `client.CosignPkToken( ...)` a requirement, why not just have one function which makes the appropriate calls based on the configuration.

> I don't think it's wise to have both Auth and OidcAuth in our API if we can help it. It violates the principle of having precisely one way to perform an action, which I would expect a library at v1.0 to adhere to.

Good call, that is exactly the plan!

`Auth` is the one call that does both or neither depending on configuration. `OidcAuth` exists for compatibility with code outside the project. Once this PR is merged, I'm going to make PRs into Docker's and SSHpiper's code to point them at `Auth` and then remove `OidcAuth`.

johncmerfeld replied:
Fair enough. I definitely don't have a good technical solution for you, and I agree that, given the constraints you've outlined, your approach makes sense!

EthanHeilman replied:
I would love a better solution, but I've been thinking about ways to do this since June 2023 and this is good enough.

johncmerfeld replied:
Agreed!

lgmugnier replied:
I know this comment will never be read, but to put in my two cents here, I don't think committing to the other projects that use us is necessary since everything is version locked

EthanHeilman commented on client/cos.go:
```diff
@@ -0,0 +1,261 @@
+// Copyright 2024 OpenPubkey
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
```
I'm putting this in the files I touch in PR to not overwrite history. At some point we will need to just hit the remaining files.

lgmugnier replied:
I'm sorry, I'm not sure what you mean by overwrite history. I'm just suggesting that it would make the best sense to do this (since its so important) in its own PR so that it doesn't clutter up this one and we can have better tracking of the changes

johncmerfeld replied:
I see Ethan's point. If we do it later it does kind of pollute the version history of an individual file and make it harder to track a set of changes.

That said, I think until we have a v1 released, anything's fair game, and it would make this PR easier to review if we didn't do it this way. I'd vote to add it to every file at a later date

johncmerfeld commented on client/providers/google.go:
```diff
@@ -60,10 +77,16 @@ func (g *GoogleOp) RequestTokens(ctx context.Context, cicHash string) (*memguard
 
-	ch := make(chan []byte)
-	chErr := make(chan error)
+	ch := make(chan []byte, 1)
+	chErr := make(chan error, 1)
```
I'm always a little wary when I see a change like this. I think you address this in a comment somewhere but is the purpose of the buffer to improve performance or to prevent a deadlock?

(I really hope it's the former!)

EthanHeilman replied:
This was a change requested by @ymarcus93 in his code review of this PR.

https://github.com/openpubkey/openpubkey/pull/68#discussion_r1444759044

I remember there being a much longer thread about this somewhere in the PR, but I can't seem to find it at the moment.  The short version is that the default in go if no size is specified is for a channel to have a buffer size of 0 elements. This makes the channel automatically block which could prevent the HTTP server from being cleaned up.  As a I understand it, the recommendation is to have channels of size 1 to prevent blocking behavior unless such blocks are an intended part of the design.

johncmerfeld reviewed (commented):
This is not a comprehensive review -- in fact I skipped over the meat of your changes because I'm still digesting them, so to speak. Nevertheless these are some questions and comments I had on the periphery that I thought you might want to chew on.

EthanHeilman commented on client/providers/github_actions.go:
```diff
@@ -130,3 +146,7 @@ func (g *GithubOp) RequestTokens(ctx context.Context, cicHash string) (*memguard
 	return client.ErrNonGQUnsupported
 }
+
+func (g *GithubOp) GetIssuer() string {
+	return githubIssuer
+}
```
Good catch! I added `GetIssuer()` in this PR and then another PR added `Issuer()` to accomplish the same thing. The other PR got merged in to main so I merged it into my branch, but I missed this when doing cleanup for that merge. I'll fix this.

EthanHeilman replied:
I'm also going to rename CosignerProvider `cos.GetIssuer()` to `cos.Issuer`

EthanHeilman commented on client/providers/mocks.go:
```diff
@@ -1,3 +1,19 @@
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
```
Yes it does violate that rule. There were changes here but the changes were removed after the merge as they were no longer needed. I'll fix.

EthanHeilman replied:
Fixed

EthanHeilman commented on pktoken/pktoken.go:
```diff
@@ -118,10 +141,13 @@ func (p *PKToken) AddSignature(token []byte, sigType SignatureType) error {
 		protected := signature.ProtectedHeaders()
-		if err := protected.Set(jws.TypeKey, string(sigType)); err != nil {
-			return err
+		if sigTypeFound, ok := protected.Get(jws.TypeKey); !ok {
+			return fmt.Errorf("required 'typ' claim not found in protected")
+		} else if sigTypeFoundStr, ok := sigTypeFound.(string); !ok {
+			return fmt.Errorf("'typ' claim in protected must be a string but was a %T", sigTypeFound)
+		} else if sigTypeFoundStr != string(sigType) {
+			return fmt.Errorf("incorrect 'typ' claim in protected, expected (%s), got (%s)", sigType, sigTypeFound)
```
That is always an excellent question to ask about code.

The unittests in this PR discovered a minor bug here. The bug was that rather than check if the `sigTyp` was correct via `protected.Get` the original code did `protected.Set` which would overwrite the original `sigType` if the `sigTyp` was incorrect. This would lead to a confusing signature verification error later in the code because `Set` was changing a signed part of the PK Token. I talked it over with Jonny if this should be a separate PR or part of this PR, but this bug can't really be triggered outside of this PR since the OP sets the `typ` for non-cosigner signatures, we decided to fix it in this PR.

As part of this bug fix, I added a regression test for the bug here:
https://github.com/openpubkey/openpubkey/pull/68/files/5e1aaeacf06e9a5ad0d0bf2cc325f18f13f13417#diff-a7210017d8c7048bfecb244430a6d961a59559c007d3354aad8993e06a406c9bR115

EthanHeilman commented on examples/ssh/opkssh/cli.go:
```diff
@@ -148,7 +164,7 @@ func authorizedKeysCommand(userArg string, typArg string, certB64Arg string, pol
 
 func createSSHCert(cxt context.Context, client *client.OpkClient, signer crypto.Signer, alg jwa.KeyAlgorithm, gqFlag bool, principals []string) ([]byte, []byte, error) {
-	pkt, err := client.OidcAuth(cxt, signer, alg, map[string]any{}, gqFlag)
+	pkt, err := client.Auth(cxt, signer, alg, map[string]any{}, gqFlag)
```
Created a issue to remove `OidcAuth` and update associated unittests

https://github.com/openpubkey/openpubkey/issues/91

EthanHeilman commented on client/oidc.go:
```diff
@@ -76,7 +92,54 @@ type OpenIdProvider interface {
+	HookHTTPSession(h http.HandlerFunc)
+}
+
 func VerifyPKToken(ctx context.Context, pkt *pktoken.PKToken, provider OpenIdProvider) error {
+	return PKTokenVerifier{
+		AllowedProviders: []OpenIdProvider{provider},
+	}.Verify(ctx, pkt)
+}
+
+type PKTokenVerifier struct {
+	AllowedProviders []OpenIdProvider
+	AllowedCosigners []CosignerProvider
+}
```
` VerifyPKToken` is my attempt to minimize the changes of this PR before internally and externally. 

For instance [OpenPubkey/signed-attestation](https://github.com/openpubkey/signed-attestation/blob/5ebc1df5bc0719b1b12dd73849efb9c7f481e2ff/opk.go#L61) uses `client.VerifyPKToken`.

We are likely to be iterating on the PKTokenVerifier over the coming months. Moving it to it's own package, improving  OP and Cosigner pre-matching logic. I don't want to have each of these changes require code churn and downstream changes, if code churn isn't needed.

Right now [OpenPubkey's golang package documentation](https://pkg.go.dev/github.com/openpubkey/openpubkey) is pretty threadbare. One of my next projects after this PR is fixing that. I want to steer people to use `client.VerifyPKToken` if they don't need the features of ` PKTokenVerifier` until it becomes more stable.

johncmerfeld commented on cosigner/authstatestore.go:
```diff
@@ -0,0 +1,188 @@
+	RedeemAuthcode(authcode string) (authState AuthState, authID string, err error)
+}
+
+// This is intended for testing purposes. The locking strategy used is not
+// particularly efficient. Anyone building a Cosigner should use the interface
+// above to replace this in-memory store with a database.
```
I think my one big ask is to move any code that this comment applies to into the examples folder. By its nature, this repo will benefit from a very clear demarcation between library code and implementations. I think there's maybe just two things that warrant the move:

  1. The implementation of `AuthStateInMemoryStore`. That would leave this file with just the interface definition, which then might be best moved to authcosigner.go
  2. Possibly the specific implementation of `AuthIDIssuer`?

But once you do that, you may conclude that some other things also count as example code.

EthanHeilman replied:
Moved `AuthStateInMemoryStore` to `cosigner/mocks` as it was originally intended to for testing and then repurposed by example MFA Cosigner. Let me know if that resolves this issue for you.

`AuthIDIssuer` is intended for production use. I wanted an atomic increment unique ID generator that is 256-bit in length and I didn't want to assume everyone will use a SQL database backend for this purpose.

johncmerfeld commented on examples/mfa/mfacosigner/mfacosigner.go:
```diff
@@ -0,0 +1,157 @@
+	"github.com/openpubkey/openpubkey/cosigner"
+)
+
+type UserKey struct {
+	Issuer string // ID Token issuer (iss)
+	Aud    string // ID Token audience (aud)
+	Sub    string // ID Token subject ID (sub)
+}
```
Since we also define this struct in the cosigner folder, it seems like **either** this should import from there, or else this whole struct is just an example and we shouldn't have the definition in cosigner.

EthanHeilman replied:
UserKey should exist in `cosigner` and be imported from there since implementations should extend it. I deleted it from MFACosigner. That struct was unused. Should be fixed

johncmerfeld commented on examples/mfa/mfacosigner/mfacosigner.go:
```diff
@@ -0,0 +1,157 @@
+		return nil, err
+	}
+
+	authCos, err := cosigner.New(signer, alg, issuer, keyID)
```
Per my comment elsewhere, this is where we should create inject the `AuthStateInMemoryStore`

EthanHeilman replied:
Should be fixed now

johncmerfeld reviewed (commented):
Still looking at a couple bigger-picture but in general this looks good to me. I'm ready to approve soon. Most important comment here is about `AuthStateInMemoryStore`.

johncmerfeld commented on examples/mfa/mfacosigner/user.go:
```diff
@@ -0,0 +1,54 @@
+	credentials []webauthn.Credential
+}
+
+var _ webauthn.User = (*user)(nil)
```
Then again, it's just example code 🤷‍♂️

EthanHeilman replied:
It is common enough in golang that it isn't worth a comment. 

For instance the golang crypto.ssh package has `var _ ...` in 202 classes. 
https://github.com/search?q=repo%3Agolang%2Fcrypto+var+_&type=code

EthanHeilman commented on client/cos.go:
```diff
@@ -0,0 +1,257 @@
+	// write to the channel. If the callbackPath handler is called twice by the
+	// user's web browser the second call will block on a channel until the cxt
+	// is marked as done.
+	ch := make(chan []byte, 1)
```
`sigCh` is definitely an improvement. Fixed!

EthanHeilman commented on client/cos.go:
```diff
@@ -0,0 +1,257 @@
+		return nil, fmt.Errorf("cosigner client hit error serializing PK Token: %w", err)
+	}
+
+	initAuthMsgJson, nonce, err := c.CreateInitAuthSig(redirectURI)
```
Good catch. Fixed!

EthanHeilman commented on client/cos.go:
```diff
@@ -0,0 +1,257 @@
+}
+
+func (c *CosignerProvider) ValidateCos(cosSig []byte, expectedNonce string, expectedRedirectURI string) error {
+	if cosSigParsed, err := jws.Parse(cosSig); err != nil {
+		return fmt.Errorf("failed to parse Cosigner signature: %w", err)
+	} else if len(cosSigParsed.Signatures()) != 1 {
+		return fmt.Errorf("the Cosigner signature does not have the correct number of signatures: %w", err)
+	} else {
+		ph := cosSigParsed.Signatures()[0].ProtectedHeaders()
+		if nonceRet, ok := ph.Get("nonce"); !ok {
+			return fmt.Errorf("nonce not set in Cosigner signature protected header")
+		} else if expectedNonce != nonceRet {
+			return fmt.Errorf("incorrect nonce set in Cosigner signature")
+		}
+		if ruriRet, ok := ph.Get("ruri"); !ok {
+			return fmt.Errorf("ruri (redirect URI) not set in Cosigner signature protected header")
+		} else if expectedRedirectURI != ruriRet {
+			return fmt.Errorf("unexpected ruri (redirect URI) set in Cosigner signature, got %s expected %s", ruriRet, expectedRedirectURI)
+		}
+		if issRet, ok := ph.Get("iss"); !ok {
+			return fmt.Errorf("iss (Cosigner Issuer) not set in Cosigner signature protected header")
+		} else if c.Issuer != issRet {
+			return fmt.Errorf("unexpected iss (Cosigner Issuer) set in Cosigner signature, expected %s", c.Issuer)
+		}
+		return nil
+	}
```
Accepted suggestion. I agree, this is easier to read because you don't have to keep track of the IF/ELSE branches

EthanHeilman commented on cosigner/authcosigner.go:
```diff
@@ -0,0 +1,117 @@
+			Signer: signer},
+		Issuer:         issuer,
+		KeyID:          keyID,
+		AuthStateStore: NewAuthStateInMemoryStore(hmacKey),
```
We do not expect developers to implement their own `AuthCosigner`, instead they should compose the `AuthCosigner` supplied with their own MFA cosigner.
The idea would be that developers would use this `AuthCosigner` and then bring their own `AuthStateStore`.

> I would expect New to take an AuthStateStore as an argument, and NewAuthStateInMemoryStore should only be invoked in examples or tests.

You are correct. `New` should definitely take a `AuthStateStore` parameter.

EthanHeilman replied:
Fixed!

EthanHeilman commented on client/providers/github_actions.go:
```diff
@@ -1,3 +1,19 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
```
Fixed!

johncmerfeld reviewed (approved):
Thanks for integrating my feedback! This is looking great. Approved.

(I'm not able to mark my comments as resolved, as I am not a committer to this project)

EthanHeilman commented:
After passing review from @johncmerfeld I recorded a final test


https://github.com/openpubkey/openpubkey/assets/274814/fcf79892-36fe-4933-9c80-252c327c3f6e
